library/alloc/src/iter.rs

```diff
@@ -0,0 +1,40 @@
+//! Allocation extensions for [`Iterator`].
+//!
+//! *[See also the Iterator trait][Iterator].*
+#![unstable(feature = "iterator_join", issue = "75638")]
+
+use crate::slice::Join;
+use crate::vec::Vec;
+
+/// Iterator extension traits that requires allocation.
+#[unstable(feature = "iterator_join", issue = "75638")]
+pub trait IteratorExt: Iterator {
+    /// Flattens an iterator into a single value with the given separator in
+    /// between.
+    ///
+    /// Combines `collect` with `join` to convert a sequence into a value
+    /// separated with the specified separator.
+    ///
+    /// Allows `.join(sep)` instead of `.collect::<Vec<_>>().join(sep)`.
+    ///
+    /// ```
+    /// #![feature(iterator_join)]
+    /// use alloc::iter::IteratorExt;
+    ///
+    /// assert_eq!(["hello", "world"].iter().copied().join(" "), "hello world");
+    /// assert_eq!([[1, 2], [3, 4]].iter().copied().join(&0), [1, 2, 0, 3, 4]);
+    /// assert_eq!([[1, 2], [3, 4]].iter().copied().join(&[0, 0][..]), [1, 2, 0, 0, 3, 4]);
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_join", issue = "75638")]
+    #[must_use = "if you really need to exhaust the iterator, consider `.for_each(drop)` instead"]
+    fn join<Separator>(self, sep: Separator) -> <[Self::Item] as Join<Separator>>::Output
```

Comment:
Sigh. [Here we go again...](https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.join) :-(

I'm *really* worried about the floodgates of future itertools breaking this trait opens up. It's bad enough that the Itertools [readme](https://github.com/rust-itertools/itertools/blob/cae2ab2331d0fc365a7094f27f22b73ca731192a/README.rst#how-to-contribute) has to beg people not to contribute methods that could reside on `Iterator`, but at least we can welcome methods that involve allocation:
> However, if your feature involves heap allocation, such as storing elements in a Vec<T>, then it can't be accepted into libcore, and you should propose it for itertools directly instead.

An `IteratorExt` trait in `alloc` blows this advice up. If such a trait is introduced, I'm not sure there's *any* situation in which we can accept useful PRs without risking future conflict with the standard libraries. Itertools is less and less an ergonomic multitool than it is a ticking timebomb that will break your builds if you have the audacity to `rustup`. [**Can we *please* fix this problem?**](https://github.com/rust-lang/rfcs/pull/2845)

Reply:
I recall that `Itertools::intersperse` also exists, how was this issue solved?

Reply:
Ugh. It's *not* solved. Stabilizing `Iterator::intersperse` *will* cause breakages for Itertools users. In that case, the fix will be as simple as upgrading one's itertools dependency to a new version *without* `Itertools::intersperse`, since the signatures and behavior of `Itertools::intersperse` and `Iterator::intersperse` is virtually identical. That's not the case with `join`, unfortunately.

Reply:
How about we do a breaking change in `itertools` to test out this `join`? Previously the team did suggest to make a change in `itertools` but the `join` is already there and is quite limited, this is similar but there is a better choice on the type of things joinable.